STM32G4: add support for G4 series, and LPUART

H-Grobben opened the pull request:
#6808 was closed due to a rename in out git repo, so I'm opening a new pull request on the new branch.
#5396: thanks for the pointer, I have tested it with the Nucleo G4 board.

H-Grobben committed:
STM32G4: Updates for STM32G4 series.

H-Grobben committed:
STM32G4: Add readme for connecting REPL uart.

H-Grobben committed:
STM32G4: Add G4 family.

H-Grobben committed:
STM32G4: Building succeded, now test if it still all works.

H-Grobben committed:
STM32G4: Set compiler opt to build succesfully.

H-Grobben committed:
STM32G4: Transfer to AEMICS repo.

H-Grobben committed:
STM: Add set_baudrate for lpuart.

dpgeorge commented:
I now have some STM32G4 hardware so can test this.

H-Grobben committed:
STM32 F4 / F7: Don't have LPUART, remove include after pyb test.

H-Grobben commented:
Ah, great, I have the Nucleo board that's included in the boards folder (G474RE), and tested it quite well. The only thing I missed was LPUART support (now included in this version, but I don't have other Nucleo boards to test if they all work).

Sorry for still adding stuff, I seem to have some troubles with the ci scripts. I fixed the pyb issue, but due to STM32lib switched to vendor (I integrated it into vendor, as it is based on the vendor version) the build now fails on some unfixed bugs in the vendor tree.

H-Grobben commented:
If you need some help with integrating into the right branches etc, please leave a note and I will try to help.

dpgeorge commented:
The stm32lib repository has now been updated to include the G4 HAL, on the branch: `work-F0-1.9.0+F4-1.16.0+F7-1.7.0+G4-1.3.0+H7-1.6.0+L0-1.11.2+L4-1.8.1+WB-1.10.0`

chrismas9 reviewed (commented):
ports/stm32/boards/stm32g474.ld

I think the fs cache is overlapping the heap and file writes will crash the middle of the heap (if used).  CCMRAM at 0x10000000 is also mapped after SRAM1 and SRAM2 and is the last part of the continuous 128KB at 0x20000000. There is only 128KB ram total comprising SRAM1 (80KB), SRAM2 (16KB) and CCMRAM (32KB).

The best layout for devices with continuous RAM is to not define RAM at 0x10000000 and just use one block at 0x20000000. Have a look at the STM32L496 linker script. Layout would be 4k fs_cache at top, then stack, then heap.

Also for devices with uniform flash sector size there is no need for FLASH_ISR. It's only needed for the F4 & F7 series to take advantage of the small sectors at the start of flash. Again look at the L496 for simpler config with FLASH first, then FLASH_FS.

H-Grobben commented:
Sorry for the delay, had some other priorities, I'll pick it up today. I've changed the setup for the G4 as @chrismas9 suggested, and the build still work on the Nucleo board. Now running the ci tests to see if it all builds.

H-Grobben committed:
STM32G4: Changed loader script for g474.

H-Grobben committed:
STM32G4: Changed loader script for g474.

H-Grobben committed:
STM: Timer callback fails. Add stop before start to fix.

H-Grobben committed:
STM32G4: Fix Timer1 callback issue.

H-Grobben committed:
STM32G4: Extint vector mapping fix.

H-Grobben commented:
I've updated with the comments or @chrismas9 and did some more testing. Found a bug in my code for timer1 callback, but also for general callback (also tested on WB series). You can't register a callback when the timer is running. Stopping before starting the timer in the callback registering function fixes this.

H-Grobben commented:
Please put this temporarily on hold, a new board with USB connection does not yet function with this port. Need to add USB support correctly.

H-Grobben committed:
STM32: Fix merge issue.

H-Grobben committed:
STM32G4: Fix for HSE enabled devices, use HSE as clock source.

H-Grobben committed:
STM32G4: Last walktrough, update with last status of master.

H-Grobben commented:
After a 'short' delay because of other priorities, we managed to gen the USB mounting working on our own dev-board. This is now incorporated in the pull request. What we fixed in the last stages:
* USB block device mounting: now capable of mounting via USB
* FDCAN bug fixes: interrupts not working correctly, now OK
* Timer fixes: saw strange behaviour after initialisation of timers. Assign {0} to structs for correct initialisation

H-Grobben committed:
STM32G4: Code formating fix.

H-Grobben commented:
To my opinion all functionality is implemented. We've tested the G4 functionality on some of our own boards, and we fixed some issues. Next to our own boards, we also added the Nucleo G474 board for everyone to use.

What can I do to help with this merge?

H-Grobben commented:
Also tested with #8057 on Nucleo_G474 and own custom board with CAN connectivity.

H-Grobben commented:
@dpgeorge / @chrismas9  I would like to get all ci checks in the green, but I don't think I could easily fix old commit messages to get them into the style of the Code Conventions (sorry my mistake...). If there is an easy way to fix them, could you point me in the right direction?

robert-hh commented:
Using `git rebase -i <before_first_commit_to_change_hash>` you can edit and change old commits. Changing the title is the easiest part, done with the `reword` action. Check the git documentation for usage and examples.

H-Grobben commented:
> Using `git rebase -i <before_first_commit_to_change_hash>` you can edit and change old commits. Changing the title is the easiest part, done with the `reword` action. Check the git documentation for usage and examples.

I have looked into this, but that doesn't seem like the proper fix: I need to reword some commits from last year, and I just merged the master. If I reword my commits, I also (can) change the commits from the master (all hash-es change from the starting point, and I can change any commit message). And that is not what I _want_ to. And I don't think the developers behind Micropython would like me to do this...

Another way is to rebase the whole G4 change into the current master, but that is a lot of additional work, checks and tests, I don't think I can persuade my boss to let me do this for some commit-message fixes.

dpgeorge commented:
Can you please rebase this on latest master?  The new stm32lib was merged with G4 HAL support.

H-Grobben committed:
STM32G4: Rebased to 30b6ce8 with final test and fixes.

H-Grobben committed:
STM32G4: CI check code formatting, and removed comment from board file.

H-Grobben commented:
> Can you please rebase this on latest master? The new stm32lib was merged with G4 HAL support.

Rebased and reword applied. Tests currently running on Nucleo G474.

dpgeorge commented on .gitmodules:
```diff
@@ -12,8 +12,8 @@
-	url = https://github.com/micropython/stm32lib
-	branch = work-F4-1.13.1+F7-1.5.0+L4-1.3.0
+	url = https://github.com/AEMICS/stm32lib
+	branch = vendor
```
This needs to be undone.

H-Grobben replied:
Done, I added the branch with G4:
work-F0-1.9.0+F4-1.16.0+F7-1.7.0+G4-1.3.0+H7-1.6.0+L0-1.11.2+L4-1.17.0+WB-1.10.0+WL-1.1.0

dpgeorge commented on ports/stm32/Makefile:
```diff
@@ -139,7 +140,7 @@ LDFLAGS += --gc-sections
 ifeq ($(DEBUG), 1)
 CFLAGS += -g -DPENDSV_DEBUG
-COPT = -Og
+COPT = -Os
```
This shouldn't be changed.

H-Grobben replied:
Ah, during debugging I changed it here... Changed it. Should I somehow revert to the base, or is a commit with the correct value OK?

dpgeorge commented on ports/stm32/boards/NUCLEO_G474RE/board_init.c:
```diff
@@ -0,0 +1,5 @@
+
+void NUCLEO_G474RE_board_early_init(void) {
+
+}
```
Unless there's a good reason to have this, since it's empty better to delete it (keeps things simpler).

H-Grobben replied:
Right, I was questioning myself to keep it for 'just in case we need it later' . So now I removed it.

dpgeorge commented on ports/stm32/boards/NUCLEO_G474RE/mpconfigboard.mk:
```diff
@@ -0,0 +1,21 @@
+#MICROPY_PY_URE = 0
+#MICROPY_PY_FRAMEBUF = 0
+#MICROPY_PY_USOCKET = 0
+#MICROPY_PY_NETWORK = 0
```
These commented out items can all be removed (and most of these don't work at the linker level).

H-Grobben replied:
Removed all the comments.

dpgeorge commented on ports/stm32/boards/NUCLEO_G474RE/mpconfigboard.mk:
```diff
@@ -0,0 +1,21 @@
+#MICROPY_PY_FRAMEBUF = 0
+#MICROPY_PY_USOCKET = 0
+#MICROPY_PY_NETWORK = 0
+FROZEN_MANIFEST = 0
```
This is not setting it correctly.   Use `FROZEN_MANIFEST =` if you don't want to freeze in any modules (but I recommend to just remove this line and pick up the default set of frozen modules).

H-Grobben replied:
Removed it.

dpgeorge commented on ports/stm32/boards/NUCLEO_G474RE/mpconfigboard.mk:
```diff
@@ -0,0 +1,21 @@
+# MCU settings
+MCU_SERIES = g4
+CMSIS_MCU = STM32G474xx
+MICROPY_FLOAT_IMPL = double
```
Does this MCU have hardware double FP?

H-Grobben replied:
Good point. It has single precision FP. Changed it.

dpgeorge commented on ports/stm32/boards/NUCLEO_G474RE/stm32g4xx_hal_conf.h:
```diff
@@ -0,0 +1,61 @@
+  * <h2><center>&copy; Copyright (c) 2019 STMicroelectronics.
+  * All rights reserved.</center></h2>
+  *
+  * This software component is licensed by ST under BSD 3-Clause license,
```
In order to not over complicate licensing, please redo this file basing it off, eg, `boards/NUCLEO_H743ZI/stm32h7xx_hal_conf.h`

H-Grobben replied:
Done.

dpgeorge commented on ports/stm32/boards/stm32g4xx_hal_conf_base.h:
```diff
@@ -0,0 +1,325 @@
+  * <h2><center>&copy; Copyright (c) 2019 STMicroelectronics.
+  * All rights reserved.</center></h2>
+  *
+  * This software component is licensed by ST under BSD 3-Clause license,
```
As above, to not over complicate licensing, please redo this file based on, eg, `boards/stm32h7xx_hal_conf_base.h`

H-Grobben replied:
Done.

dpgeorge commented on ports/stm32/dma.c:
```diff
@@ -1108,6 +1327,12 @@ void dma_nohal_init(const dma_descr_t *descr, uint32_t config) {
+    #elif defined(STM32G4)
+    // descr = { DMA1_Channel4, DMA_REQUEST_UART3_TX, dma_id_3,  &dma_init_struct_spi_i2c };
+    // #error Hier moet ik nog ff naar kijken hoe we dit aanpassen aan de G4 serie
+    // in L4 wordt hier channel geselecteerd, in G4 ws via DMAMUX (net als H7?)
```
If this comment is warranted, please convert it to English :smile:

H-Grobben replied:
Removed it, was a comment from the start. DMA is kind of different in this G4 compared to most ST processors, as it has a DMA and DMAMUX peripheral, and I had to learn how ST implemented the DMA.

dpgeorge commented on ports/stm32/pin_defs_stm32.h:
```diff
@@ -128,6 +128,14 @@ enum {
+#define CAN2 FDCAN2
+#define GPIO_AF9_CAN1 GPIO_AF9_FDCAN1
+#define GPIO_AF9_CAN2 GPIO_AF9_FDCAN2
+#endif
```
Is this still needed, now that your FDCAN PR was merged?

dpgeorge commented:
Thanks @H-Grobben for the effort on this, there is a lot of work here.

I left some comments which are hopefully not too much work to respond to.  If your time is short I can work them in during merge.  Just let me know.

H-Grobben committed:
STM32G4: Implemented review comments.

H-Grobben committed:
STM32G4: Rebase to last head.

H-Grobben commented:
Testing done. Still have issues with the ci_code_formatting_run on the ports/stm32/make-stmconst.py file. When running locally, some py files get extra newlines. But they are rejected by the github ci process.

H-Grobben committed:
STM32G4: RNG did not produce random nr due to clock issue.

H-Grobben commented:
Tests showed that the RNG did not produce fully random numbers. This was caused by a clock issue on the RNG peripheral. Now the HSI48 is enabled and selected for RNG (and for future PR also for USB).

dpgeorge commented:
Thanks for responding to the review comments.  It looks like the CI failures are now gone, which is good!

There are many commits in this PR and I will squash them all down into one during merge.

dpgeorge commented:
Squashed and merged in 8f68e26f79ee325669fb7dc7bbd092fccd5b038f with the following minor changes (that did not change the compiled firmware image):
- removed README.md because the LPUART now works as REPL
- moved `defined(STM32G4)`'s to sorted order with respect to MCU series (eg F4, G4, H7, L0)
- fixed second heading row in new af.csv file
- factored new code in make-pins.py
- removed unnecessary changes to pyb_can.c
- split change to .gitmodules to a separate commit

Thanks for the effort on this!

dpgeorge commented on ports/stm32/mpconfigboard_common.h:
```diff
@@ -394,6 +394,15 @@
+#define PYB_EXTI_NUM_VECTORS (42) // up to 42 event/interrupt requests: 28 configurable lines, 14 direct lines
+#define MICROPY_HW_MAX_I2C (3)
+#define MICROPY_HW_MAX_TIMER (20) // TIM 1-8, 20
+#define MICROPY_HW_MAX_UART (5) // uart 1 - 5 + lpuart1
```
@H-Grobben I'm wondering if this should be 4, not 5.  LPUART1 is excluded from this count, I think.  Eg there is the following:
```
struct _pyb_uart_obj_t *pyb_uart_obj_all[MICROPY_HW_MAX_UART + MICROPY_HW_MAX_LPUART];
```

H-Grobben replied:
It depends on the specific type. There are types with 3 USARTS (and LPUART) only, and the added Nucleo type (G474Rx) has 3 USARTS and 2 UARTS (and LPUART). I was in doubt on the number I should include here, but my decision was the number of peripherals in the Nucleo board, so that's why it was 5.

dpgeorge replied:
Ah, I see, thanks!  So it is correct as is, it should be 5 because that is the maximum possible on the G4 series.
